compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1097,76 +1096,201 @@ private BType inferTableMemberType(List<BType> memTypes, BType expType) {
+                BRecordType bMemType = (BRecordType) memType;
+                if (bMemType.sealed || bMemType.fields.containsKey(fieldName)) {
+                    continue;
                 }
+
+                BType restFieldType = bMemType.restFieldType;
+                types.add(restFieldType);
+            }
+
+            BField resultantField = createFieldWithType(fields.get(0), types);
+            boolean isOptional = hasOptionalFields(fields) || fields.size() != memTypesSize;
+
+            if (isOptional) {
+                resultantField.symbol.flags = Flags.OPTIONAL;
+            } else if (keySpecifierFieldNames.contains(fieldName)) {
+                resultantField.symbol.flags = Flags.REQUIRED | Flags.READONLY;
+            } else {
+                resultantField.symbol.flags = Flags.REQUIRED;
             }
+
+            inferredFields.add(resultantField);
         }
 
-        Set<BField> commonFieldSet = new LinkedHashSet<>(allFieldSet);
-        for (BType memType : memTypes) {
-            commonFieldSet.retainAll(((BRecordType) memType).fields.values());
+        return createTableConstraintRecordType(inferredFields, restFieldTypes, tableConstructorExpr.pos);
+    }
+
+    /**
+     * Create a new {@code BField} out of existing {@code BField}, while changing its type.
+     * The new type is derived from the given list of bTypes.
+     *
+     * @param field  - existing {@code BField}
+     * @param bTypes - list of bTypes
+     * @return a {@code BField}
+     */
+    private BField createFieldWithType(BField field, List<BType> bTypes) {
+        LinkedHashSet<BType> effectiveTypes = getEffectiveMemberTypes(bTypes);
+
+        BType newType;
+        if (effectiveTypes.size() == 1) {
+            newType = effectiveTypes.iterator().next();
+        } else {
+            newType = BUnionType.create(null, effectiveTypes);
         }
 
-        List<String> requiredFieldNames = new ArrayList<>();
-        if (keySpecifier != null) {
-            for (IdentifierNode identifierNode : keySpecifier.fieldNameIdentifierList) {
-                requiredFieldNames.add(((BLangIdentifier) identifierNode).value);
-                keySpecifierFieldNames.add(((BLangIdentifier) identifierNode).value);
+        BVarSymbol originalSymbol = field.symbol;
+        BVarSymbol fieldSymbol = new BVarSymbol(originalSymbol.flags, originalSymbol.name, originalSymbol.pkgID,
+                newType, originalSymbol.owner, originalSymbol.pos, VIRTUAL);
+
+        return new BField(field.name, field.pos, fieldSymbol);
+    }
+
+    private LinkedHashSet<BType> getEffectiveMemberTypes(List<BType> bTypeList) {
+        LinkedHashSet<BType> bTypes = new LinkedHashSet<>(bTypeList);
+
+        if (bTypeList.size() == 1) {
+            return bTypes;
+        }
+
+        LinkedHashSet<BType> eBTypes = new LinkedHashSet<>(bTypes.size());
+        addEffectiveMemberTypes(eBTypes, bTypes);
+
+        eBTypes = filterRedundantAnyAnydataSubtypes(eBTypes);
```

Comment:
We can just use the `isAssignable` checks instead and filter out a type, right?

I think it would be something like
- org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker#getRepresentativeBroadType
- or the type logic in org.wso2.ballerinalang.compiler.semantics.analyzer.Types#inferRecordFieldType

Basically filtering out only the super types from the set while still representing the same value space.

That way we don't have to special case types and all types including user-defined types that are the same as any/anydata will also be covered.